File: server/src/Feature/Login/LoginService.hs
module Feature.Login.LoginService
  ( login
  , generateJwtToken
  )
where

import Protolude
import Data.Aeson
import Control.Monad.Except
import Feature.User.User
import Feature.User.UserRepoClass
import Feature.Login.LoginResult
import qualified Infrastructure.Utils.Crypto as Crypto
import qualified Infrastructure.Secrets as Secrets
import qualified Web.JWT as JWT

data Login = Login
  { loginUsername :: Text
  , loginPassword :: Text
  } deriving Generic

instance FromJSON Login where
  parseJSON =
    withObject "login" $ \o -> Login <$> o .: "username" <*> o .: "password"

login :: UserRepo m => LByteString -> m (Either LoginError Text)
login rawBody = runExceptT $ do
  body <- liftEither $ parseBody rawBody
  user <- ExceptT $ findUserByCredentials body
  return $ generateJwtToken (userUsername user)

parseBody :: LByteString -> Either LoginError Login
parseBody rawBody = maybeToRight InvalidRequest $ decode rawBody

findUserByCredentials :: UserRepo m => Login -> m (Either LoginError User)
findUserByCredentials req = do
  userInDb <- mfilter isPasswordValid <$> findUserByUsername (loginUsername req)

  return $ maybeToRight UserNotFound $ userInDb
 where
  isPasswordValid :: User -> Bool
  isPasswordValid user =
    Crypto.validate (userPassword user) (loginPassword req)

generateJwtToken :: Text -> Text
generateJwtToken username = JWT.encodeSigned key mempty cs
 where
  cs = mempty
    { JWT.iss = JWT.stringOrURI "listeo"
    , JWT.sub = JWT.stringOrURI username
    }
  key = JWT.hmacSecret Secrets.jwtSecret

File: server-ts/src/features/login/LoginService.ts
import jwt from 'jsonwebtoken'
import { EitherAsync } from 'purify-ts/EitherAsync'
import { Codec, string, GetInterface } from 'purify-ts/Codec'
import { findUserByUsername } from '../user/UserRepo'
import { Pool } from 'pg'
import { User } from '../user/User'
import { compareSync } from 'bcrypt'
import { Either } from 'purify-ts/Either'
import { jwtSecret } from '../../infrastructure/Secrets'

export enum LoginError {
  UserNotFound = 'UserNotFound',
  InvalidRequest = 'InvalidRequest'
}

const LoginBody = Codec.interface({
  username: string,
  password: string
})

type LoginBody = GetInterface<typeof LoginBody>

export const login = (
  env: Env,
  rawBody: unknown
): EitherAsync<LoginError, string> =>
  EitherAsync(async ({ liftEither, fromPromise }) => {
    const body = await liftEither(parseBody(rawBody))
    const user = await fromPromise(findUserByCredentials(env.pool, body))
    return generateJwt(user.username)
  })

const parseBody = (rawBody: unknown): Either<LoginError, LoginBody> =>
  LoginBody.decode(rawBody).mapLeft(_ => LoginError.InvalidRequest)

const findUserByCredentials = (
  pool: Pool,
  body: LoginBody
): Promise<Either<LoginError, User>> => {
  const isPasswordValid = (user: User): boolean =>
    compareSync(user.password, body.password)

  return findUserByUsername(body.username, pool).then(maybeUser =>
    maybeUser.filter(isPasswordValid).toEither(LoginError.UserNotFound)
  )
}

const generateJwt = (username: string): string =>
  jwt.sign(username, jwtSecret, { algorithm: 'RS256', issuer: 'listeo' })
